Save LAEA grid

abooton opened the pull request:
**Description**
Saving of cubes on the Lambert Azimuthal Equal Area (LAEA) projection is added.
The LAEA template is described here: https://codes.ecmwf.int/grib/format/grib2/templates/3/140/

It is very similar to the Lambert Conformal projection specification in template_30, and so the coding was based on that.
https://codes.ecmwf.int/grib/format/grib2/templates/3/30/

**Associated Issue**
https://github.com/SciTools/iris-grib/issues/344

pp-mo commented:
I just realised that we have an outstanding previous request to do this (plus another thing) ... #295.
The discussion there seems to have stalled, but we should probably choose one over the other somehow.

trexfeathers commented on iris_grib/_save_rules.py:
```diff
@@ -649,6 +650,65 @@ def grid_definition_template_30(cube, grib):
+                      central_lon / _DEFAULT_DEGREES_UNITS)
+    eccodes.codes_set(grib, 'resolutionAndComponentFlags',
+                      0x1 << _RESOLUTION_AND_COMPONENTS_GRID_WINDS_BIT)
+
```
... plus a test

trexfeathers reviewed (commented):
Thanks @abooton! I've compared this to the templates (thanks for the links), the Lambert Conformal code, and to #295; the changes make sense.

I have a few questions/suggestions below.

trexfeathers commented:
@lbdreyer @abooton any thoughts on progressing this?

trexfeathers commented:
Happy to close this if no longer important

abooton commented:
Sorry - I get so much GitHub spam I had not realised it had review comments on it.  This is still important.

abooton commented on iris_grib/_save_rules.py:
```diff
@@ -649,6 +650,65 @@ def grid_definition_template_30(cube, grib):
+                      int(np.round(first_y / _DEFAULT_DEGREES_UNITS)))
+    eccodes.codes_set(grib, "longitudeOfFirstGridPoint",
+                      int(np.round(first_x / _DEFAULT_DEGREES_UNITS)))
+    eccodes.codes_set(grib, 'standardParallel',
+                      central_lat / _DEFAULT_DEGREES_UNITS)
+    eccodes.codes_set(grib, 'centralLongitude',
+                      central_lon / _DEFAULT_DEGREES_UNITS)
```
https://codes.ecmwf.int/grib/format/grib2/templates/3/140/

abooton commented on iris_grib/_save_rules.py:
```diff
@@ -649,6 +650,65 @@ def grid_definition_template_30(cube, grib):
+    central_lon = cs.longitude_of_projection_origin % 360
+    central_lat = cs.latitude_of_projection_origin
+
+    eccodes.codes_set(grib, "latitudeOfFirstGridPoint",
+                      int(np.round(first_y / _DEFAULT_DEGREES_UNITS)))
+    eccodes.codes_set(grib, "longitudeOfFirstGridPoint",
+                      int(np.round(first_x / _DEFAULT_DEGREES_UNITS)))
+    eccodes.codes_set(grib, 'standardParallel',
+                      central_lat / _DEFAULT_DEGREES_UNITS)
+    eccodes.codes_set(grib, 'centralLongitude',
+                      central_lon / _DEFAULT_DEGREES_UNITS)
+    eccodes.codes_set(grib, 'resolutionAndComponentFlags',
+                      0x1 << _RESOLUTION_AND_COMPONENTS_GRID_WINDS_BIT)
```
I guessed based on the other code, particularly that of loading.  I'm not a GRIB expert - I was hoping this would be checked as part of the review process!

trexfeathers replied:
OK, sorry for putting the onus on you. Most Iris-GRIB PRs I have reviewed have been from GRIB experts by coincidence so I have developed a bad habit.

Here's what I have after further investigation. **TLDR you've already covered all the bases**:

- Octets 15 to 30: handled by this line:
  https://github.com/SciTools/iris-grib/blob/70d35ef0073f27c22345be07d3eca0301c6873e5/iris_grib/_save_rules.py#L688
- Octets 31 to 38: `numberOfPointsAlongXAxis` and `numberOfPointsAlongYAxis` are aliased to `Nx` and `Ny`, which are handled by this line:
  https://github.com/SciTools/iris-grib/blob/70d35ef0073f27c22345be07d3eca0301c6873e5/iris_grib/_save_rules.py#L688
- Octets 39 to 55: you are already handling
- Octets 56 to 63: `xDirectionGridLengthInMillimetres` and `yDirectionGridLengthInMillimetres` are alised to `Dx` and `Dy`, which are handled by these lines:
  https://github.com/SciTools/iris-grib/blob/70d35ef0073f27c22345be07d3eca0301c6873e5/iris_grib/_save_rules.py#L685-L686
- Octet 64: handled by this line:
  https://github.com/SciTools/iris-grib/blob/70d35ef0073f27c22345be07d3eca0301c6873e5/iris_grib/_save_rules.py#L688

trexfeathers commented:
Posting this outside any review comments to give me a common place to reference.

I tracked down [`template.3.140.def` in eccodes](https://github.com/ecmwf/eccodes/blob/a1f17da1086d4b0e0abc8c194023a84a3de5c0d6/definitions/grib2/template.3.140.def), from which I generated this:

|**Octets**|**Key**                                        |**Type**     |**Content**                                            |
|------|-------------------------------------------|---------|---------------------------------------------------|
|15    |shapeOfTheEarth                            |codetable|Shape of the Earth (see Code table 3.2)            |
|16    |scaleFactorOfRadiusOfSphericalEarth        |unsigned |Scale factor of radius of spherical Earth          |
|17-20 |scaledValueOfRadiusOfSphericalEarth        |unsigned |Scaled value of radius of spherical Earth          |
|21    |scaleFactorOfMajorAxisOfOblateSpheroidEarth|unsigned |Scale factor of major axis of oblate spheroid Earth|
|22-25 |scaledValueOfMajorAxisOfOblateSpheroidEarth|unsigned |Scaled value of major axis of oblate spheroid Earth|
|26    |scaleFactorOfMinorAxisOfOblateSpheroidEarth|unsigned |Scale factor of minor axis of oblate spheroid Earth|
|27-30 |scaledValueOfMinorAxisOfOblateSpheroidEarth|unsigned |Scaled value of minor axis of oblate spheroid Earth|
|31-34 |numberOfPointsAlongXAxis                   |unsigned |Nx - numberOfPointsAlongXAxis                      |
|35-38 |numberOfPointsAlongYAxis                   |unsigned |Ny - numberOfPointsAlongYAxis                      |
|39-42 |latitudeOfFirstGridPoint                   |signed   |La1 - latitudeOfFirstGridPoint                     |
|43-46 |longitudeOfFirstGridPoint                  |signed   |Lo1 - longitudeOfFirstGridPoint                    |
|47-50 |standardParallelInMicrodegrees             |signed   |Standard Parallel In Microdegrees                  |
|51-54 |centralLongitudeInMicrodegrees             |signed   |Central Longitude In Microdegrees                  |
|55    |resolutionAndComponentFlags                |codeflag |Resolution and component flags (see Flag table 3.3)|
|56-59 |xDirectionGridLengthInMillimetres          |unsigned |Dx - X-direction grid length in millimetres        |
|60-63 |yDirectionGridLengthInMillimetres          |unsigned |Dy - Y-direction grid length in millimetres        |
|64    |scanningMode                               |codeflag |Scanning mode (see Flag table 3.4)                 |

abooton commented:
I have:
* Renamed `standardParallel` to `standardParallelInMicrodegrees` (in `_save_rules.py` and in the unit tests)
* Renamed `centralLongitude` to `centralLongitudeInMicrodegrees` (in `_save_rules.py` and in the unit tests)
* Added the raising the exception if the false_easting and false_northing are not zero. (I can see this can be configured when creating the LAEA coord-system, but there isn't the slot in the GRIB to save it, hence the need to raise the exception).
* I am happy that the lat and lon are set in degrees in the test `_make_test_cube` and are coded to be in micro-degrees in the `test__template_specifics`

trexfeathers commented on iris_grib/tests/unit/save_rules/test_grid_definition_template_140.py:
```diff
@@ -0,0 +1,133 @@
+        cs = self._default_coord_system(false_easting=false_easting,
+                                        false_northing=false_northing)
+        test_cube = self._make_test_cube(cs=cs)
+        msg = ('non zero false easting ({:.2f}) or '
+               'non zero false northing ({:.2f})'
+               '; unsupported by GRIB Template 3.140'
+               '.').format(false_easting, false_northing)
```
```suggestion
        msg = (r'non zero false easting \(\d*\.\d{2}\) or '
               r'non zero false northing \(\d*\.\d{2}\)'
               r'; unsupported by GRIB Template 3\.140'
               r'\.')
```

The evils of regex mean that brackets and periods need escaping. As part of this it seemed easier to not attempt to capture the actual numbers in the failure test - these would then need more escape characters! I instead asserted for any 2-point decimal number.

trexfeathers commented:
The docs build failure is not a concern for this pull request - #365

trexfeathers commented:
Thanks very much for your contribution @abooton
